napari/layers/shapes/_shape_list.py

```diff
@@ -353,6 +352,170 @@ def add(
             # Set z_order
             self._update_z_order()
 
+    def add_multiple(
+        self,
+        shapes,
+        face_colors=None,
+        edge_colors=None,
+        z_refresh=True,
+    ):
+        """Add multiple shapes at once (faster than adding them one by one)
+
+        Parameters
+        ----------
+        shapes : iterable of Shape
+            Each Shape must be a subclass of Shape, one of "{'Line', 'Rectangle',
+            'Ellipse', 'Path', 'Polygon'}"
+        face_colors : iterable of face_color
+        edge_colors : iterable of edge_color
+        z_refresh : bool
+            If set to true, the mesh elements are reindexed with the new z order.
+            When shape_index is provided, z_refresh will be overwritten to false,
+            as the z indices will not change.
+            When adding a batch of shapes, set to false  and then call
+            ShapesList._update_z_order() once at the end.
+
+        TODO: Currently shares a lot of code with `add()`, with the
+        difference being that `add()` supports inserting shapes at a specific
+        `shape_index`, whereas `add_multiple` will append them as a full batch
+        """
+
+        def _make_index(length, shape_index, cval=0):
+            """Same but faster than `np.repeat([[shape_index, cval]], length, axis=0)`"""
+            index = np.empty((length, 2), np.int32)
+            index.fill(cval)
+            index[:, 0] = shape_index
+            return index
```

Comment:
Why is this function nested?

Reply:
since it's an arcane idiosyncratic helper function specific to the enclosing function :)

Reply:
ie no reason :joy: We should pull it out. One of the NASA programming principles for bug-free code is functions should be 25 lines or less. Nested functions make that ~impossible to achieve. Obviously there's other bits that need compressing in this function but this is a trivial win.

Reply:
I would agree in general. But the semantic scope of `make_index` is clearly confined to the enclosing function (i.e. useless to the rest of the code) and thus pulling it out would imho actually decrease maintainability